rook/scripts/funcs.sh

```diff
@@ -0,0 +1,400 @@
+    fi
+    if [[ "$LOGGING" == "--logging-enabled" ]]; then
+      logJob "${1}" "${2}" "${4}"
+    fi
+}
+
+# Args:
+#   1. namespace
+#   2. name
+#   3. cluster
+function logJob {
+    writeLog "${1}" "${2}" "Job" "${3}"
+    writeEvent "${1}" "${2}" "Job" "${3}"
+    writeEvent "${1}" "${2}" "Pod" "${3}"
+}
+
+LOGSFOLDER="logs"
+EVENTSFOLDER="events"
+
+# This function writes logs to file for specified <kind>
+# Args:
+#   1. namespace
+#   2. name
+#   3. kind
+#   4. cluster
+function writeLog {
+    if [[ -z "$PIPELINE" ]]; then
+        return
+    fi
+
+    NAMESPACE=$1
+    NAME=$2
+    KIND=$3
+    CLUSTER=$4
+    NAMES=$(kubectl get "$KIND" -n "$NAMESPACE" -o custom-columns=NAME:.metadata.name | grep "$NAME" | tail -n +1)
+    mapfile -t NAMESLIST <<< "$NAMES"
+
+    mkdir -p "./$LOGSFOLDER/$CLUSTER/$KIND/$NAMESPACE"
+    for NAME in "${NAMESLIST[@]}"
+    do
+        FILE="./$LOGSFOLDER/$CLUSTER/$KIND/$NAMESPACE/$NAME.log"
+        if [[ ! -f "$FILE" ]]; then
+            touch "$FILE"
+            kubectl -n "$NAMESPACE" logs "$KIND"/"$NAME" --all-containers=true > "$FILE" 2>&1
+        fi
+    done
+}
+
+# This function writes events to file for specified <kind>
+# Args:
+#   1. namespace
+#   2. name
+#   3. kind
+#   4. cluster
+function writeEvent {
+    if [[ -z "$PIPELINE" ]]; then
+        return
+    fi
+
+    NAMESPACE=$1
+    NAME=$2
+    KIND=$3
+    CLUSTER=$4
+    NAMES=$(kubectl get "$KIND" -n "$NAMESPACE" -o custom-columns=NAME:.metadata.name | grep "$NAME" | tail -n +1)
+    mapfile -t NAMESLIST <<< "$NAMES"
+
+    mkdir -p "./$EVENTSFOLDER/$CLUSTER/$KIND/$NAMESPACE"
+    for NAME in "${NAMESLIST[@]}"
+    do
+        FILE="./$EVENTSFOLDER/$CLUSTER/$KIND/$NAMESPACE/$NAME.event"
+        if [[ ! -f "$FILE" ]]; then
+            touch "$FILE"
+            DATA=$(kubectl get event -n "${NAMESPACE}" --field-selector involvedObject.kind="${KIND}",involvedObject.name="${NAME}" -o json)
+            MESSAGES=$(echo "${DATA}" | jq -r '.items | map(.message) | .[]')
+            echo "$MESSAGES" > "$FILE"
+        fi
+    done
+}
+
+# Args:
+#   1. Name of cephcluster
+function testCephCluster {
+  jsonData=$(kubectl -n rook-ceph get cephclusters.ceph.rook.io "${1}" -ojson)
+  health=$(echo "${jsonData}" | jq -r '.status.ceph.health')
+  phase=$(echo "${jsonData}" | jq -r '.status.phase')
+  if [[ ! $health = "HEALTH_OK" ]]; then
+    echo "Cluster health is not ok: $health ❌"; FAILURES=$((FAILURES+1))
+    DEBUG_OUTPUT+=$(echo "${jsonData}" | jq '.status.ceph.details')
+  else
+    echo "Cluster health is ok ✔"
+  fi
+  if [[ ! $phase = "Ready" ]]; then
+    echo "Cluster phase is ready: $phase ❌"; FAILURES=$((FAILURES+1))
+  else
+    echo "Cluster phase is ready ✔"; SUCCESSES=$((SUCCESSES+1))
+  fi
+}
+
+function testNumberOfOSDs {
+  numberOfOSDs=$(kubectl -n rook-ceph get deployments.apps -o=jsonpath='{.items[*].metadata.name}' |  xargs -n 1 | grep -c rook-ceph-osd)
```

Comment:
Why do we need the xargs part?
```suggestion
  numberOfOSDs=$(kubectl -n rook-ceph get deployments.apps -o=jsonpath='{.items[*].metadata.name}' |  grep -c rook-ceph-osd)
```